src/objective/rank_obj.cu

```diff
@@ -356,52 +427,40 @@ struct NDCGLambdaWeightComputer {
     }
 
    private:
-     const float *dsorted_labels_{nullptr};  // Labels sorted within a group
-     const uint32_t *dorig_pos_{nullptr};  // Original indices of the labels before they are sorted
-     const uint32_t *dgroups_{nullptr};  // The group indices
      const float *dgroup_dcg_ptr_{nullptr};  // Start address of the group DCG values
-     // Where can a prediction for a label be found in the original array, when they are sorted
-     uint32_t *dindexable_sorted_preds_pos_ptr_{nullptr};
   };
 
   NDCGLambdaWeightComputer(const bst_float *dpreds,
-                           uint32_t pred_size,
+                           const bst_float *dlabels,
                            const SegmentSorter<float> &segment_label_sorter)
-    : dgroup_dcg_(segment_label_sorter.NumGroups()),
-      dindexable_sorted_preds_pos_(pred_size),
-      weight_multiplier_(segment_label_sorter.Items(),
-                         segment_label_sorter.OriginalPositions(),
-                         segment_label_sorter.GroupIndices(),
-                         dgroup_dcg_.data().get(),
-                         dindexable_sorted_preds_pos_.data().get()) {
-    // Sort the predictions first and get the sorted position
-    SegmentSorter<float> segment_prediction_sorter;
-    segment_prediction_sorter.SortItems(dpreds, pred_size, segment_label_sorter.GroupSegments());
-
-    this->CreateIndexableSortedPredictionPositions(segment_prediction_sorter.OriginalPositions());
-
-    // Compute each group's DCG concurrently
-    // Set the values to be the group indices first so that the predicate knows which
-    // group it is dealing with
-    thrust::sequence(dgroup_dcg_.begin(), dgroup_dcg_.end());
-
-    // TODO(sriramch): parallelize across all elements, if possible
-    // Transform each group - the predictate computes the group's DCG
-    thrust::transform(dgroup_dcg_.begin(), dgroup_dcg_.end(),
-                      dgroup_dcg_.begin(),
-                      ComputeGroupDCG(segment_label_sorter.Items(),
-                                      segment_label_sorter.GroupIndices()));
-  }
-
-  inline NDCGLambdaWeightMultiplier GetWeightMultiplier() const { return weight_multiplier_; }
-  inline const dh::caching_device_vector<uint32_t> &GetSortedPredPos() const {
-    return dindexable_sorted_preds_pos_;
+    : IndexablePredictionSorter(dpreds, segment_label_sorter),
+      dgroup_dcg_(segment_label_sorter.GetNumGroups(), 0.0f),
+      weight_multiplier_(segment_label_sorter, *this) {
+    const auto &group_segments = segment_label_sorter.GetGroupSegments();
+
+    // Compute each elements DCG values and reduce them across groups concurrently.
+    auto end_range =
+      thrust::reduce_by_key(group_segments.begin(), group_segments.end(),
```

Comment:
`thrust::reduce_by_key` is notoriously slow, it has to dynamically work out how many keys exist then allocate memory accordingly. If you know the number of keys in advance [cub::DeviceSegmentedReduce](https://nvlabs.github.io/cub/structcub_1_1_device_segmented_reduce.html) has much better performance.

If this section of code is not performance critical then using thrust is fine.

Reply:
i instrumented the code to compute the overheads and here is what i found:

- for the small mslr dataset, the fusing of transform and reduction (which is done here) took 18.2ms for 100 rounds, which is roughly ~ 182 microseconds per tree building round. out of this, half the time is spent on the transform operation - which is computing the item's dcg weight. hence, the overhead of reduction is ~ 90 microseconds per call

- for the large dataset it is ~ 400 microseconds for the transform and reduction of which reduction took ~ half the time

Reply:
So you are saying the performance impact is negligible, is that correct?